staging/src/k8s.io/apiserver/pkg/authentication/token/cache/cached_token_authenticator_test.go

```diff
@@ -172,6 +175,99 @@ func BenchmarkKeyFunc(b *testing.B) {
 	})
 }
 
+func TestSharedLookup(t *testing.T) {
+	t.Run("actually shared", func(t *testing.T) {
+		var lookups uint32
+		c := make(chan struct{})
+		a := New(authenticator.TokenFunc(func(ctx context.Context, token string) (*authenticator.Response, bool, error) {
+			<-c
+			atomic.AddUint32(&lookups, 1)
+			return &authenticator.Response{User: &user.DefaultInfo{Name: "chewbaca"}}, true, nil
+		}), true, time.Minute, 0)
+
+		var wg sync.WaitGroup
+		for i := 0; i < 10; i++ {
+			wg.Add(1)
+			go func() {
+				defer wg.Done()
+				a.AuthenticateToken(context.Background(), "")
+			}()
+		}
+		// no good way to make sure that all the callers are queued so we sleep.
```

Comment:
Do you actually need this with the wait group?

Reply:
I need to wait for the goroutines to block inside of AuthenticateToken which isn't something that the wg provides. I picked the sleep time by running this test 10k times without a flake.

Reply:
There is probably some approach with a channel + drain that would avoid the sleep.

Since CI can be a lot slower than our machines, can we at least use a sleep time in the order of seconds?

Reply:
Ok, I will fix. Just being lazy.

Reply:
I bumped to a second. I could throw in a second waitgroup that gets decremented when the AuthenticateToken goroutines start, but I'd still need a sleep to deschedule the test goroutine.

Reply:
Since you're caching for a minute, I guess this must have been demonstrating the race I was talking about.

Reply:
I was expecting the race to be really rare but if we have to sleep a whole second to make this non-racy for just 10 callers I'm wondering if maybe we should think about doing something about it?

Reply:
The race is basically fixed by double checking the cache after entering the singleflight. I suppose there's not a good way to test this because it's an internal detail of the singleflight. However, I've left the test around because it's not flaky and doesn't seem like much of a burden.